### PR TITLE
Speculative attempt to fix recursion errors in tests

### PR DESCRIPTION
**Pull request by DRMacIver**
This is an attempt at #799. It's not a very good attempt, but I think it will work.

You will note the absence of tests. This is because based on a bunch of investigation just now, this is a *really* hard problem to reproduce. I've not been able to get a test to reliably reproduce the problem despite some trying.

Based on this I'm 99% certain that the reason for this is that this is not an infinite recursion error, just a lots recursion error, and the difficulty with reproducing comes from there being less on the stack when I try, but working around that is fiddly.

Based on further investigation of what is on the stack at the time of this happening, I came to the conclusion that the reason this was happening is that in some highly specific circumstances we call `do_draw` directly rather than going through `draw`. This means that we're putting a bunch of extra stack frames in there without incrementing the depth counter on the data object. This means that since #756 raised the max depth, we sometimes put enough on the stack that we overflow the recursion limit.

So this PR just implements plan "don't do that then" and replaces the `do_draw` calls with draw calls. I've been running the flaky test from #799 in a loop for a while now and haven't seen any failures of it, so I'm reasonably confident that this works despite the absence of testing.

### CODE REVIEW

**Comment by DRMacIver** [comment]
I propose after merging this we wait a week and close the linked issue (it's really hard not using the word f i x next to the issue number for fear github will pick it up and c l o s e it. Hilarious unintended consequences) if we haven't seen this fail on CI since.

**Comment by Zac-HD** [comment]
This looks good to me, but a design question: if we have code that avoids creating RecursionErrors, why do we have a MAX_DEPTH constant at all?

**Comment by DRMacIver** [comment]
> if we have code that avoids creating RecursionErrors, why do we have a MAX_DEPTH constant at all?

The `MAX_DEPTH` constant *is* our code for avoiding creating RecursionErrors. We don't look at the current stack depth because that's unreliable - it would give different answers e.g. when generating the test and replaying the test.

File: RELEASE.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+RELEASE_TYPE: patch
+
+This release will fix some extremely specific circumstances that probably have
```
Comment (Zac-HD) [inline RELEASE.rst]: No future tense - "this release fixes ..." by the time anyone reads it!

(I generally describe new of changed behaviour in present tense, and fixed bugs in past tense)

**Comment by Zac-HD** [comment]
Ah, right - essentially we're betting that we never use more than about `sys.getrecursionlimit() / MAX_DEPTH` stack frames per known-to-Hypothesis recursion, with a bit of padding for initial depth etc.  Makes sense.

With the comment above, LGTM.

**Comment by DRMacIver** [comment]
> Ah, right - essentially we're betting that we never use more than about sys.getrecursionlimit() / MAX_DEPTH stack frames per known-to-Hypothesis recursion, with a bit of padding for initial depth etc. 

Yeah. And by default sys.getrecursionlimit() is 1000, so we're reasonably comfortably under that as long as we don't fool the checker like this. It could cause problems if you call Hypothesis from deep in the stack, but recommended fix in that case is to use `setrecursionlimit`...

**Pull request merged**